File: plugins/sulis-execution/sdk/typescript/src/resources/work_package.ts
import { resultPayload } from '../helpers.js';
import type {
  AsyncSubprocessTransport,
  SubprocessTransport,
  TransportConfig,
} from '../transport.js';
import type {
  WorkPackageAppendEvidenceResult,
  WorkPackageReadMetadataResult,
} from '../types.js';

/**
 * Work Package resource — wraps `wpx-wp` subcommands.
 *
 * Underlying CLI: `wpx-wp`. The SDK exposes `readMetadata` instead of
 * the CLI's `read-frontmatter` (which leaks markdown-author jargon).
 */

const BINARY = 'wpx-wp';

function common(config: TransportConfig): Record<string, unknown> {
  return { project: config.project, repo_root: config.repoRoot };
}

export class WorkPackageResource {
  constructor(
    private readonly transport: SubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  /** Read frontmatter metadata from a WP file. CLI subcommand: read-frontmatter. */
  readMetadata(params: {
    wp: string;
    field: string;
  }): WorkPackageReadMetadataResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'read-frontmatter', {
        ...common(this.config),
        ...params,
      }),
    ) as unknown as WorkPackageReadMetadataResult;
  }

  /** Append `## Acceptance Evidence` to a WP file. */
  appendEvidence(params: {
    wp: string;
    evidence_json: string;
  }): WorkPackageAppendEvidenceResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'append-evidence', {
        ...common(this.config),
        ...params,
      }),
    ) as unknown as WorkPackageAppendEvidenceResult;
  }
}

export class AsyncWorkPackageResource {
  constructor(
    private readonly transport: AsyncSubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  async readMetadata(params: {
    wp: string;
    field: string;
  }): Promise<WorkPackageReadMetadataResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'read-frontmatter', {
        ...common(this.config),
        ...params,
      }),
    ) as unknown as WorkPackageReadMetadataResult;
  }

  async appendEvidence(params: {
    wp: string;
    evidence_json: string;
  }): Promise<WorkPackageAppendEvidenceResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'append-evidence', {
        ...common(this.config),
        ...params,
      }),
    ) as unknown as WorkPackageAppendEvidenceResult;
  }
}

File: plugins/sulis-execution/sdk/python/sulis_execution/resources/work_package.py
"""Work Package resource — wraps `wpx-wp` subcommands.

Read metadata from a Work Package file; append acceptance evidence to it.

Underlying CLI: `wpx-wp`. The SDK wraps that under cleaner names —
`read_metadata` instead of the CLI's `read-frontmatter` (which leaks
markdown-author jargon).
"""
from __future__ import annotations

from sulis_execution._helpers import _result_payload
from sulis_execution.transport import (
    AsyncSubprocessTransport,
    SubprocessTransport,
    TransportConfig,
)
from sulis_execution.types import (
    WorkPackageAppendEvidenceResult,
    WorkPackageReadMetadataResult,
)

BINARY = "wpx-wp"


def _common(config: TransportConfig) -> dict:
    return {"project": config.project, "repo_root": str(config.repo_root)}


class WorkPackageResource:
    def __init__(
        self, transport: SubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    def read_metadata(
        self, *, wp: str, field: str
    ) -> WorkPackageReadMetadataResult:
        """Read frontmatter metadata from a WP file.

        Returns one field's value, or the whole frontmatter dict if
        field is "*". Underlying CLI subcommand: `read-frontmatter`.
        """
        params = {**_common(self._config), "wp": wp, "field": field}
        envelope = self._transport.invoke(BINARY, "read-frontmatter", params)
        return WorkPackageReadMetadataResult.model_validate(_result_payload(envelope))

    def append_evidence(
        self, *, wp: str, evidence_json: str
    ) -> WorkPackageAppendEvidenceResult:
        """Append the `## Acceptance Evidence` section to a WP file."""
        params = {**_common(self._config), "wp": wp, "evidence_json": evidence_json}
        envelope = self._transport.invoke(BINARY, "append-evidence", params)
        return WorkPackageAppendEvidenceResult.model_validate(_result_payload(envelope))


class AsyncWorkPackageResource:
    def __init__(
        self, transport: AsyncSubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    async def read_metadata(
        self, *, wp: str, field: str
    ) -> WorkPackageReadMetadataResult:
        params = {**_common(self._config), "wp": wp, "field": field}
        envelope = await self._transport.invoke(BINARY, "read-frontmatter", params)
        return WorkPackageReadMetadataResult.model_validate(_result_payload(envelope))

    async def append_evidence(
        self, *, wp: str, evidence_json: str
    ) -> WorkPackageAppendEvidenceResult:
        params = {**_common(self._config), "wp": wp, "evidence_json": evidence_json}
        envelope = await self._transport.invoke(BINARY, "append-evidence", params)
        return WorkPackageAppendEvidenceResult.model_validate(_result_payload(envelope))
